pkg/controllers/hub/trafficmanagerbackend/controller.go

```diff
@@ -285,6 +294,39 @@ func (r *Reconciler) validateAzureTrafficManagerProfile(ctx context.Context, bac
 	return &getRes.Profile, nil
 }
 
+// validateServiceImportAndCleanupEndpointsIfInvalid returns not nil serviceImport when the serviceImport is valid.
+func (r *Reconciler) validateServiceImportAndCleanupEndpointsIfInvalid(ctx context.Context, backend *fleetnetv1alpha1.TrafficManagerBackend, azureProfile *armtrafficmanager.Profile) (*fleetnetv1alpha1.ServiceImport, error) {
+	backendKObj := klog.KObj(backend)
+	var cond metav1.Condition
+	serviceImport := &fleetnetv1alpha1.ServiceImport{}
+	if getServiceImportErr := r.Client.Get(ctx, types.NamespacedName{Name: backend.Spec.Backend.Name, Namespace: backend.Namespace}, serviceImport); getServiceImportErr != nil {
+		if apierrors.IsNotFound(getServiceImportErr) {
+			klog.V(2).InfoS("NotFound serviceImport and starting deleting any stale endpoints", "trafficManagerBackend", backendKObj, "serviceImport", backend.Spec.Backend.Name)
+			if err := r.cleanupEndpoints(ctx, backend, azureProfile); err != nil {
+				klog.ErrorS(err, "Failed to delete stale endpoints for an invalid serviceImport", "trafficManagerBackend", backendKObj, "serviceImport", backend.Spec.Backend.Name)
+				return nil, err
+			}
+			cond = metav1.Condition{
+				Type:               string(fleetnetv1alpha1.TrafficManagerBackendConditionAccepted),
+				Status:             metav1.ConditionFalse,
+				ObservedGeneration: backend.Generation,
+				Reason:             string(fleetnetv1alpha1.TrafficManagerBackendReasonInvalid),
+				Message:            fmt.Sprintf("ServiceImport %q is not found", backend.Spec.Backend.Name),
+			}
+			meta.SetStatusCondition(&backend.Status.Conditions, cond)
+			backend.Status.Endpoints = []fleetnetv1alpha1.TrafficManagerEndpointStatus{} // none of the endpoints are accepted by the TrafficManager
+			return nil, r.updateTrafficManagerBackendStatus(ctx, backend)
+		}
+		klog.ErrorS(getServiceImportErr, "Failed to get serviceImport", "trafficManagerBackend", backendKObj, "serviceImport", backend.Spec.Backend.Name)
+		setUnknownCondition(backend, fmt.Sprintf("Failed to get the serviceImport %q: %v", backend.Spec.Profile.Name, getServiceImportErr))
```

Comment:
Here we remove all the endpoints in the backend status but we leave orphan endpoints on the ATM resource. I suspect this will cause resource leak.

Reply:
we don't rely on the serviceImport Status to cleanup the endpoints. Here we will retry and rebuild the desired endpoints (in my next PR).